\begin{abstract}

Over the past few years advances in genetic sequencing hardware have introduced
the concept of massively parallel DNA sequencing, reducing both the time and cost
involved in performing genetic analysis. However, these
"next-generation" processes are complicated and open to error, thus
quality control is an essential step to assure confidence in any downstream
analyses carried out.

This project investigates the accuracy of a range of decision tree classifiers
for a series of quality control data sets and shows it is possible to
generate minimal yet accurate models, creating trees which closely resemble the
behaviour of an already existing QC system.

This project also introduces \textbf{Frontier}, a Python package which provides
users with interfaces for the reading, storage and retrieval of large machine
learning data sets; \textbf{Goldilocks}, a Python class which can be used
to find a region on a genome that expresses a desired density of variants across
different studies; and contributions to internal tools used as the Sanger
Institute and existing widely used open-source bioinformatics software.

\end{abstract}
